server/public/model/plugin_push_notification.go

```diff
@@ -0,0 +1,12 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package model
+
+// PluginPushNotification is sent to the plugin when a push notification is going to be sent (via the
+// NotificationWillBePushed hook).
+type PluginPushNotification struct {
+	Post    *Post
+	Channel *Channel
+	UserID  string
+}
```

Comment:
Why do we send only a subset of the fields instead of just sending all of `model.PushNotification`?

Reply:
If we extend that to include what we need for the current use case (`PostType` and `ChannelType` at the least I am thinking) then it should work fine.

Reply:
As Claudio mentioned, we would have to add the fields we require. But I think the bigger issue is that `model.PushNotification` is created at a later point in the push notification chain. The issue I think is: do we want to allow plugins to override server settings such as `IdLoadedNotification` (as I mentioned above in the summary)? I don't think we should, and therefore we need to put the hook somewhere before `model.PushNotification` is created. I chose this point as it's a single point that all push notifications pass.

Reply:
I do admit that this is awkward, and its because we're hooking into a push notification early in the processing chain. Perhaps we can hook after the `model.PushNotification` has been created, but then we'd have to do new processing after it's returned from the hook (to strip the message and maybe other fields, we'd have to be extra careful). I didn't want to go that route because of the complication.

Reply:
Personally I am not worried about the capabilities concerns given plugins (as designed) can do pretty much whatever they want already so spending effort on limiting what they can do seems like a drop in the ocean sort of deal. 

Of course I still think we should make it as simple as possible for implementers to avoid common mistakes through easy to understand interfaces.